coriolis/providers/replicator.py

```diff
@@ -283,6 +283,114 @@ def init_replicator(self):
+        after the execution of the attachment operation.
+
+        param disk_id: str(): the 'disk_id' of the info from self._volumes_info
+            for the disk which shall be attached by `attachf()`.
+        param attachf: fun(): argument-less function which attaches the disk.
+            The function should perform any waits required to reasonably expect
+            the worker OS to have noticed the new disk, or reboot the worker VM
+            and re-run `init_replicator()` if deemed necessary.
+        param previous_disks_status: dict(): previous status of the disks as
+            returned by `get_current_disks_status()`.
+
+        return: dict(): returns the volumes_info associated to the new disk.
+        """
+        # check if volume with given ID is present in self._volumes_info:
+        matching_vols = [
+            vol for vol in self._volumes_info
+            if vol['disk_id'] == disk_id]
+        if not matching_vols:
+            raise exception.CoriolisException(
+                "No information regarding volume with ID '%s'. "
+                "Cannot track its attachment." % disk_id)
+        elif len(matching_vols) > 2:
+            raise exception.CoriolisException(
+                "Multiple volumes info with ID '%s' found: %s" % (
+                    disk_id, matching_vols))
+        vol_info = matching_vols[0]
+
+        # get/refresh current device paths:
+        if not previous_disks_status:
+            previous_disks_status = self._cli.get_status()
+        previous_device_paths = [
+            dev['device-path'] for dev in previous_disks_status]
+
+        # run attachment function and get new device paths:
+        attachf()
+        new_disks_status = self._cli.get_status()
+        new_device_paths = [dev['device-path'] for dev in new_disks_status]
+
+        # check for missing/multiple new device paths:
+        missing_device_paths = (
+            set(previous_device_paths) - set(new_device_paths))
+        if missing_device_paths:
+            LOG.warn(
+                "The following devices from the previous disk state qeury are "
+                "no longer detected: %s", [
+                    dev for dev in previous_disks_status
+                    if dev['device-path'] in missing_device_paths])
+
+        new_device_paths = set(new_device_paths) - set(previous_device_paths)
+        if not new_device_paths:
+            raise exception.CoriolisException(
+                "No new device paths have appeared after volume attachment of "
+                "disk with ID: %s" % disk_id)
+        elif len(new_device_paths) > 2:
+            raise exception.CoriolisException(
+                "Multiple device paths have appeared after attachment of disk "
+                "with ID %s: %s" % (
+                    disk_id,
+                    [dev for dev in new_disks_status
+                     if dev['device-path'] in new_device_paths]))
+
+        new_dev = [
```

Comment:
At this point you should have only one disk in ```new_device_paths```, considering you are attaching one disk at a time, or at least that's what the ```attachf()``` function should do.

Wouldn't it be easier to just assign new_device_path to ```vol_info['disk_path']```?

Reply:
These checks are all here so as to ensure `attachf()` doesn't accidentally attach more than one disk (which is likely to happen considering most current plugins implement a single `attach_disks` method, so a lazy implementation which reuses that will get caught here)

Reply:
Yes. Understood. But at this point, you already passed the above tests where you check if ```len(new_device_paths) > 1```. It should be safe to trust that we have only 1 element in that set.

Reply:
You are correct, I got a bit blind-sighted, fixing now.